cli/server.go

```diff
@@ -425,75 +453,117 @@ func server() *cobra.Command {
+			// the user about what's going on and handle errors more
+			// explicitly.
+
 			_, err = daemon.SdNotify(false, daemon.SdNotifyStopping)
 			if err != nil {
-				return xerrors.Errorf("notify systemd: %w", err)
+				cmd.Printf("Notify systemd failed: %s", err)
 			}
-			_, _ = fmt.Fprintln(cmd.OutOrStdout(), cliui.Styles.Bold.Render(
-				"Interrupt caught, gracefully exiting. Use ctrl+\\ to force quit"))
 
-			for _, provisionerDaemon := range provisionerDaemons {
-				if verbose {
-					cmd.Println("Shutting down provisioner daemon...")
-				}
-				err = provisionerDaemon.Shutdown(cmd.Context())
-				if err != nil {
-					cmd.PrintErrf("Failed to shutdown provisioner daemon: %s\n", err)
-					continue
-				}
-				err = provisionerDaemon.Close()
-				if err != nil {
-					return xerrors.Errorf("close provisioner daemon: %w", err)
-				}
-				if verbose {
-					cmd.Println("Gracefully shut down provisioner daemon!")
-				}
+			// Stop accepting new connections without interrupting
+			// in-flight requests, give in-flight requests 5 seconds to
+			// complete.
+			cmd.Println("Shutting down API server...")
+			err = shutdownWithTimeout(server, 5*time.Second)
+			if err != nil {
+				cmd.Printf("API server shutdown took longer than 5s: %s", err)
+			} else {
+				cmd.Printf("Gracefully shut down API server\n")
 			}
+			// Cancel any remaining in-flight requests.
+			shutdownConns()
+
+			// Shut down provisioners before waiting for WebSockets
+			// connections to close.
+			var wg sync.WaitGroup
+			for i, provisionerDaemon := range provisionerDaemons {
+				id := i + 1
+				provisionerDaemon := provisionerDaemon
+				wg.Add(1)
+				go func() {
+					defer wg.Done()
+
+					if verbose {
+						cmd.Printf("Shutting down provisioner daemon %d...\n", id)
+					}
+					err := shutdownWithTimeout(provisionerDaemon, 5*time.Second)
+					if err != nil {
+						cmd.PrintErrf("Failed to shutdown provisioner daemon %d: %s\n", id, err)
+						return
+					}
+					err = provisionerDaemon.Close()
+					if err != nil {
+						cmd.PrintErrf("Close provisioner daemon %d: %s\n", id, err)
+						return
+					}
+					if verbose {
+						cmd.Printf("Gracefully shut down provisioner daemon %d\n", id)
+					}
+				}()
+			}
+			wg.Wait()
+
+			cmd.Println("Waiting for WebSocket connections to close...")
+			_ = coderAPI.Close()
+			cmd.Println("Done wainting for WebSocket connections")
 
+			// Close tunnel after we no longer have in-flight connections.
 			if tunnel {
 				cmd.Println("Waiting for tunnel to close...")
 				closeTunnel()
-				<-devTunnelErrChan
+				<-devTunnelErr
+				cmd.Println("Done waiting for tunnel")
 			}
 
 			// Ensures a last report can be sent before exit!
 			options.Telemetry.Close()
-			cmd.Println("Waiting for WebSocket connections to close...")
-			shutdownConns()
-			coderAPI.Close()
-			return nil
+
+			// Trigger context cancellation for any remaining services.
+			cancel()
+
+			return exitErr
```

Comment:
We'll always be returning an error now. There's really no way to cleanly shut down the server (e.g. click shut down in WebUI), so this makes sense. And interrupt for instance usually returns non-zero exit code.

The only downside is that we will print `context cancelled` and `see --help` at the end. I kind of want us to have a cobra post-exec for cli that does `os.Exit(n)` is the context is cancelled. Thoughts?